### PR TITLE
feat(health): Add custom health check for KServe Inference Service resources

### PR DESCRIPTION
**Pull request by rachitchauhan43**
## Context
This PR adds the custom health check for KServe Inference Service Resource

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: resource_customizations/serving.kserve.io/InferenceService/health.lua [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      end
+      health_status.message = health_status.message .. msg .. "\n"
+      if condition.status == "False" then
+        status_false = status_false + 1
+      else
+        status_unknown = status_unknown + 1
+      end
+    end
+  end
+  if status_true == 3 and status_false == 0 and status_unknown == 0 then
```
Comment (crenshaw-dev) [inline resource_customizations/serving.kserve.io/InferenceService/health.lua:28]: Is `status_true == 3` necessary and sufficient? I see there are four condition types which may increment status_true: IngressReady, PredictorConfigurationReady, PredictorReady, PredictorRouteReady. Do we need all 5, or is some subset sufficient?

Comment (rachitchauhan43) [reply resource_customizations/serving.kserve.io/InferenceService/health.lua]: We need all 5 to be true. Added that.

File: resource_customizations/serving.kserve.io/InferenceService/health.lua [hunk 2]
```diff
@@ -0,0 +1,40 @@
+health_status = {}
+health_status.status = "Progressing"
+health_status.message = "Waiting for status update."
+if obj.status ~= nil and obj.status.conditions ~= nil then
+  status_true = 0
+  status_false = 0
+  status_unknown = 0
+  health_status.message = ""
+  for i, condition in pairs(obj.status.conditions) do
+    if condition.status == "True" and (condition.type == "IngressReady" or condition.type == "PredictorConfigurationReady" or condition.type == "PredictorReady" or condition.type == "PredictorRouteReady") then
```
Comment (crenshaw-dev) [inline resource_customizations/serving.kserve.io/InferenceService/health.lua:10]:
```suggestion
    if condition.status == "True" and (condition.type == "IngressReady" or condition.type == "PredictorConfigurationReady" or condition.type == "PredictorReady" or condition.type == "PredictorRouteReady" or condition.type == "Ready") then
```

File: resource_customizations/serving.kserve.io/InferenceService/health_test.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+tests:
+#- healthStatus:
+#    status: Progressing
+#    message: "PredictorConfigurationReady is Unknown\nPredictorReady is Unknown, since RevisionMissing. Configuration \"hello-world-predictor-default\" is waiting for a Revision to become ready.\nPredictorRouteReady is Unknown, since RevisionMissing. Configuration \"hello-world-predictor-default\" is waiting for a Revision to become ready.\nReady is Unknown, since RevisionMissing. Configuration \"hello-world-predictor-default\" is waiting for a Revision to become ready.\n"
+#  inputPath: testdata/progressing.yaml
+#- healthStatus:
+#    status: Degraded
+#    message: "IngressReady is False, since Predictor ingress not created.\nPredictorConfigurationReady is False, since RevisionFailed. Revision \"helloworld-00002\" failed with message: Container failed with: container exited with no error.\nPredictorReady is False, since RevisionFailed. Revision \"helloworld-00002\" failed with message: Container failed with: container exited with no error.\nReady is False, since Predictor ingress not created.\n"
+#  inputPath: testdata/degraded.yaml
```
Comment (crenshaw-dev) [inline resource_customizations/serving.kserve.io/InferenceService/health_test.yaml:9]: Wanna uncomment these and see if we get past tests?

Comment (rachitchauhan43) [reply resource_customizations/serving.kserve.io/InferenceService/health_test.yaml:9]: Yes. Just testing one by one. Once this healthy one passes I will uncomment others too.

**Review by crenshaw-dev — APPROVED** [review]
Thanks, @rachitchauhan43!

**Pull request merged**